src/main/java/games/chess/ChessGameState.java

```diff
@@ -0,0 +1,414 @@
+                    case KNIGHT:
+                    case BISHOP:
+                        playerScore += 3;
+                        break;
+                    case ROOK:
+                        playerScore += 5;
+                        break;
+                    case QUEEN:
+                        playerScore += 9;
+                        break;
+                    case KING:
+                        playerScore += 0;
+                        break;
+                }
+            }
+            //Reward check
+            if (isInCheck(1-playerId)) {
+                playerScore += 1;
+            }
+            //Penalize oppenent king movement choices
+            int opponentKingActions = computeAvailableActionsKing(getKingPosition(1-playerId)[0], getKingPosition(1-playerId)[1], 1-playerId).size();
+            playerScore -= opponentKingActions * 0.2; // TODO: Tune this value to be more or less punishing
+
+            return (playerScore - opponentScore)/100; // TODO: Normalize this value to be between 0 and 1
+        } else {
+            // The game finished, we can instead return the actual result of the game for the given player.
+            return getPlayerResults()[playerId].value;
+        }
+    }
+
+    @Override
+    public double getGameScore(int playerId) {
+        return 0;
+    }
+
+    @Override
+    protected boolean _equals(Object o) {
+        return o instanceof ChessGameState that &&
+                super.equals(o) &&
+                this.halfMoveClock == that.halfMoveClock &&
+                this.whitePieces.equals(that.whitePieces) &&
+                this.blackPieces.equals(that.blackPieces) &&
+                this.gameStateCounts.equals(that.gameStateCounts) &&
+                this.board.equals(that.board);
+    }
+    
+    @Override
+    public int hashCode() {
+        return Objects.hash(super.hashCode(), halfMoveClock, whitePieces, blackPieces, gameStateCounts, board);
+    }
+
+    public ChessBoard getBoard() {
+        return board;
+    }
+
+    public void setPiece(int x, int y, ChessPiece piece) {
+        board.setPiece(x, y, piece);
+        if (piece != null) {
+            if (piece.getOwnerId() == 0) {
+                whitePieces.add(piece);
+            } else if (piece.getOwnerId() == 1) {
+                blackPieces.add(piece);
+            } else {
+                throw new IllegalArgumentException("Invalid player ID: " + piece.getOwnerId());
+            }
+        }
+    }
+    public ChessPiece getPiece(int x, int y) {
+        return board.getPiece(x, y);
+    }
+    public List<ChessPiece> getPlayerPieces(int playerId) {
+        if (playerId == 0) {
+            return whitePieces;
+        } else if (playerId == 1) {
+            return blackPieces;
+        } else {
+            throw new IllegalArgumentException("Invalid player ID: " + playerId);
+        }
+    }
+
+    public void deletePiece(ChessPiece piece) {
+        if (piece == null) {
```

Comment:
This is potentially dangerous as it can easily lead to bugs happening invisibly and being very difficult to track down. 
If code calls deletePiece() then they are expecting a piece to be present...if it isn't then a much better approach is to throw an Error (IllegalArgument, say).
This will highlight the bug as early as possible and make it much easier to fix